### DISCUSSION TITLE
Refraction Issue

### DISCUSSION THREAD

**Discussion by rpwh** [RayTracing/raytracing.github.io#1087] [Debugging]
Sorry for posting again but I was wondering if anyone has any idea what's going on with my dielectric material/refraction. Here is a picture of what's happening:
![refraction_error](https://user-images.githubusercontent.com/22357922/197404732-164b5073-7b07-4236-a474-ff5be85bd5b8.png)
I'm not sure why the area inside the dielectric sphere's aren't being inverted horizontally and why the reflection from the metal sphere is appearing in the left sphere.

https://github.com/rpwh/ray-tracing

**Comment by NicolasGuillaumeSoulie** [comment]
There's a mistake in your dot product implementation. 
```Rust
// It should be
pub fn dot(&self, other: &Vec3) -> f32 {
    self.x * other.x + self.y * other.y + self.z * other.z
}
// But you wrote
pub fn dot(&self, other: &Vec3) -> f32 {
    self.x * other.x + self.y * other.y + self.z * self.z
} __________________________________________________^^^
```
![image](https://user-images.githubusercontent.com/48185211/199853176-f8eb4f7e-15d1-4f50-be2e-863d0cc33710.png)
Once fixed the refraction seems almost fine, but you also forgot to return black in ```fn ray_color()``` if ```rec.mat.scatter()``` return 0. Because of that your metal shine on the edges.
|![image](https://user-images.githubusercontent.com/48185211/199853387-122e0dc3-e050-4381-a790-4bac17d82d8a.png)|![image](https://user-images.githubusercontent.com/48185211/199855697-3861752b-42f0-4aa2-81cd-f2b32faf9fb9.png)|
|:--:|:--:|
|dot product fixed|dot product + ray_color fixed |


![image](https://user-images.githubusercontent.com/48185211/199854457-5fed6f42-86cb-4c99-9e5b-d2d8a8f102a8.png)

**Reply by NicolasGuillaumeSoulie** [reply]
And I just realised that my implementation (that was supposed to be completed) also lacks that black return yet I never had that glow...

**Reply by rpwh** [reply]
This is awesome! Thanks for taking the time to look through my code and find the fix!

**Reply by NicolasGuillaumeSoulie** [reply]
No problem. 
Your bug just made me curious (although I suspected it was vector based) and since I'm also doing it in Rust I thought it could be a learning experience.
